usr/src/uts/armv8/rpi4/io/bcm2711_pcie/bcm2711_pcie.c

```diff
@@ -306,6 +403,51 @@ bcm2711_pcie_attach(dev_info_t *dip, ddi_attach_cmd_t cmd)
 
 	VERIFY3U(softc->bc_base, !=, 0);
 
+	bcm2711_pcie_reset_controller(softc);
+
+	uint32_t hw_rev = bcm2711_pcie_read_reg(softc,
+	    BCM2711_REG_CONTROLLER_HW_REV) & 0xffff;
+
+	dev_err(dip, CE_NOTE, "hardware revision: 0x%x.", hw_rev);
+
+	bcm2711_pcie_enable_controller(softc);
+
+	/*
+	 * wait for controller startup
+	 */
+	for (uint_t i = 0;; i++) {
+		uint32_t bstate =
+		    bcm2711_pcie_read_reg(softc, BCM2711_REG_BRIDGE_STATE);
+
+		if ((bstate & 0x30) == 0x30)
+			break;
+
+		/*
+		 * The controller failed to start in 100 ms
+		 */
+		if (i >= 100) {
+			dev_err(dip, CE_WARN, "controller failed to start");
+			return (DDI_FAILURE);
+		}
+
+		drv_usecwait(1000);
+	}
+
+	uint32_t link_state =
+	    bcm2711_pcie_read_reg(softc, BCM2711_REG_BRIDGE_LINK_STATE) >> 0x10;
+
+	if (link_state == 0) {
+		dev_err(dip, CE_WARN, "controller started but link is not up");
+		return (DDI_FAILURE);
+	}
+
+	dev_err(dip, CE_NOTE, "reported link speed: %s",
+	    bcm2711_pcie_link_state_string(link_state));
+
+	/*
```

Comment:
OpenBSD set the window and say it's because EFI broke it, I don't think we need to _yet_ but we probably will.

Reply:
Yeah, if it's ok for you I'd like to keep the comment so we won't forget in case we need to do something.